rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2215 @@
+
+[`include_bytes`]: https://doc.rust-lang.org/core/macro.include_bytes.html
+
+The [`include_bytes`] macro statically includes the contents of another file as a byte array, producing a value of type `&'static [u8; N]`. Often, these bytes correspond to structured data and a `transmute` must be used to deserialize the bytes into useful form; e.g.:
+```rust
+pub fn recognize(input: &Matrix<f64, U1, U784>) -> usize
+{
+    static RAW_WEIGHT : &'static [u8; 62_720] = include_bytes!("/weight.bin");
+
+    static RAW_BIAS : &'static [u8; 80] = include_bytes!("/bias.bin");
+
+    let WEIGHT: &Matrix<f64, U784, U10> = unsafe{ mem::transmute(RAW_WEIGHT) };
+
+    let BIAS: &Matrix<f64, U1, U10> = unsafe{ mem::transmute(RAW_BIAS) };
+
+    network::recognize(input, WEIGHT, BIAS)
+}
+```
+
+This RFC proposes mechanisms that would eliminate `unsafe` in this example. We revisit this motivating use-case with our proposed mechanisms near the end of this RFC, [*here*][future-possibility-include_data].
+
+
+## SIMD
+***TODO:*** https://internals.rust-lang.org/t/pre-rfc-frombits-intobits/7071
+
+
+## `Atomic<T>`
+Rust defines a dozen `Atomic*` types (`AtomicBool`, `AtomicI8`, `AtomicI16`, `AtomicI32`, `AtomicI64`, `AtomicIsize`, `AtomicPtr`, `AtomicU8`, `AtomicU16`, `AtomicU32`, `AtomicU64`, and `AtomicUsize`).
+
+This set is large—a distinct `Atomic*` type is required for each primitive type—but incomplete. If one wants atomic operations on their own type, they must define a wrapper around an existing `Atomic*` type of appropriate size and validity, then transmute at API boundaries; e.g.:
+
+```rust
+#[repr(u8)]
+enum Trilean {
+    False,
+    True,
+    Unknown,
+}
+
+#[repr(transparent)]
+pub AtomicTrilean(AtomicU8);
+
+impl AtomicTrilean {
+
+    pub const fn new(v: Trilean) -> Self {
+        AtomicTrilean(
+          AtomicU8::new(
+            unsafe { mem::transmute(v) }
+          ))
+    }
+
+    ...
+}
+```
+
+The mechanisms proposed by this RFC would eliminate this pattern, permitting truly generic `Atomic` types. We revisit this motivating use-case with these proposed mechanisms near the end of this RFC, [*here*][future-possibility-generic-atomics].
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Terminology & Concepts
+
+### 📖 Transmutation
+**Transmutation** is the act of reinterpreting the bytes corresponding to a value of one type as if they corresponded to a different type.
+
+### 📖 Soundness
+A transmutation is ***sound*** if the mere act of transmuting a value from one type to another is not compiler undefined behavior.
+
+### 📖 Safety
+A sound transmutation is ***safe*** if *using* the transmuted value cannot violate memory safety.
+
+### 📖 Stability
+A safe transmutation is ***stable*** if the authors of the source type and destination types have indicated that the layouts of those types is part of their libraries' stability guarantees.
+
+## Concepts in Depth
```

Comment:
Full disclosure: If this RFC were accepted and fully implemented tonight, some of the examples in this section still wouldn't quite work *as-written* due to an existing inference issue; e.g.:
```rust
let _ : i8 = i8::transmute_from(0u8);
//              ^^^^^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteFrom`
let _ : i8 = 0u8.transmute_into();
//               ^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteInto`
```
https://github.com/rust-lang/rust/issues/27336 might be related to what's going on here. In these situations, this syntax may be used instead:
```rust
let _ : i8 = TransmuteFrom::<_>::transmute_from(0u8);
let _ : i8 = TransmuteInto::<_>::transmute_into(0u8);
```
Given that solution, I'm surprised that *this* **doesn't** work:
```rust
let _ : i8 = TransmuteFrom::transmute_from(0u8);
//           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteFrom`
let _ : i8 = TransmuteInto::transmute_into(0u8);
//           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteInto`
```
...because I would think that `Trait::method` ought to be equivalent to `Trait::<_>::method`.

---

I don't know *if* or *where* to file this issue in the RFC. It's not a drawback of the RFC per se, and it's possible the issue will be resolved by the time this RFC is stabilized.